feat: AUTOSIGN_GPG_LOCAL_FPR: selection of local signer

robbat2 opened the pull request:
Support selecting which local secret key should be used to sign other keys.

Via environment variable AUTOSIGN_GPG_LOCAL_FPR.

Under the hood, this uses gpg options: --local-user, --default-key, --trusted-key with distinct trustdb files per key.

It also supports revocation with the same configurable key.

The test suite is significantly enhanced to test revocation as well as
multiple keys.

Reference: https://bugs.gentoo.org/834329
Signed-off-by: Robin H. Johnson <robbat2@gentoo.org>

robbat2 committed:
feat: AUTOSIGN_GPG_LOCAL_FPR: selection of local signer

Support selecting which local secret key should be used to sign other
keys.

Via environment variable AUTOSIGN_GPG_LOCAL_FPR.

Under the hood, this uses gpg options: --local-user, --default-key,
--trusted-key with distinct trustdb files per key.

It also supports revocation with the same configurable key.

The test suite is significantly enhanced to test revocation as well as
multiple keys.

Reference: https://bugs.gentoo.org/834329
Signed-off-by: Robin H. Johnson <robbat2@gentoo.org>

mgorny reviewed (changes requested):
This doesn't change the logic used to verify whether the uid has been signed already, so I honestly doubt it works as expected.

robbat2 commented:
@mgorny this is ready for you to review again; I also improved the test suite to test revocation, which was not covered before.

The testing is not perfect, it would be nicer to test with a special test LDAP input and test keyring input, so we could be sure of exact outcomes.

robbat2 committed:
feat: use future time to detect upcoming sig expiry

This is as good as we can do until upstream implements something
directly in GPG to list soon-to-expire signatures (similar to the recent
key_expires_d option).

Signed-off-by: Robin H. Johnson <robbat2@gentoo.org>

mgorny commented on autosign.bash:
```diff
@@ -48,83 +66,141 @@ get_ldap() {
+	done <"${f}"
+}
+
+get_ldap_raw() {
```
I'd find it more readable if you mobrf dependent functions declared before their use.

mgorny commented on autosign.bash:
```diff
@@ -48,83 +66,141 @@ get_ldap() {
+}
+
+get_ldap_raw() {
+	h="$(hostname --fqdn)"
```
`local`

mgorny commented on autosign.bash:
```diff
@@ -48,83 +66,141 @@ get_ldap() {
+	h="$(hostname --fqdn)"
+	# This needs some weird quoting because SSH.
+	if [[ "${h%.gentoo.org}" == "${h}" ]]; then
+		ssh dev.gentoo.org "ldapsearch -Z -D '' -LLL '${AUTOSIGN_FILTER}' gpgfingerprint"
```
`${AUTOSIGN_FILTER@Q}` should take care of quoting it properly.

mgorny commented on autosign.bash:
```diff
@@ -48,83 +66,141 @@ get_ldap() {
+	# This needs some weird quoting because SSH.
+	if [[ "${h%.gentoo.org}" == "${h}" ]]; then
+		ssh dev.gentoo.org "ldapsearch -Z -D '' -LLL '${AUTOSIGN_FILTER}' gpgfingerprint"
+		rc=$?
```
You don't need to save it, it won't be overwritten over returning from `if`.

mgorny commented on autosign.bash:
```diff
@@ -1,25 +1,43 @@
+	get_ldap_raw >"${f}"
 	while read l; do
-		case ${l} in
+		case "${l}" in
```
`case` doesn't require quoting.

mgorny commented on autosign.bash:
```diff
@@ -48,83 +66,141 @@ get_ldap() {
+	local -a GPGREC
+	local l keyid='' fpr=''
+	local uid_validity='' uid_string='' uid_email=''
+	#local sig_validity= sig_class= sig_issuer=
```
What's the purpose of the commented out code?

mgorny commented on autosign.bash:
```diff
@@ -48,83 +66,141 @@ get_ldap() {
 # Get UID-fingerprint list of all currently trusted keys.
 get_signed_keys() {
-	local l keyid= fpr= trust uid email
+	local -a GPGREC
```
Why are you naming locals uppercase?

mgorny commented on autosign.bash:
```diff
@@ -48,83 +66,141 @@ get_ldap() {
+		# TODO: as a seperate pass, apply a C decode to convert the above back.
+		IFS=':' read -ra GPGREC <<< "$l"
+
+		# Be a terrible programmer: the docs have fields numbered from 1
```
Or you could just put a dummy word before `${l}`.

mgorny commented on autosign.bash:
```diff
@@ -1,25 +1,43 @@
-	local l uid= fpr system
+	local l uid='' fpr='' system=''
+	local f='ldapraw.txt'
+	get_ldap_raw >"${f}"
```
Space after `>`.

mgorny commented on autosign.bash:
```diff
@@ -48,83 +66,141 @@ get_ldap() {
+get_ldap_raw() {
+	h="$(hostname --fqdn)"
+	# This needs some weird quoting because SSH.
+	if [[ "${h%.gentoo.org}" == "${h}" ]]; then
```
In bash we do `[[ ${h} != *.gentoo.org ]]`.

mgorny commented on autosign.bash:
```diff
@@ -48,83 +66,141 @@ get_ldap() {
 revoke_sig() {
 	local key=${1}
 	local uid=${2}
+	local _gpgcmd=( "${gpgcmd[@]}" )
+	_gpgcmd+=( "${trust_args[@]}" )
```
Why are you underscoring the variable? Just use a readable name.

mgorny commented on autosign.bash:
```diff
@@ -48,83 +66,141 @@ get_ldap() {
-		send "save\n"
-		expect eof
-	EOF
+	"${_gpgcmd[@]}" \
```
Does this give the correct revocation reason?

mgorny commented on autosign.bash:
```diff
@@ -134,30 +210,51 @@ sign_key() {
 					uids+=( "=${uid}" )
 				# if there are revoked UIDs, they may collide
-				[[ ${trust} == [er] ]] && need_full=1
+				[[ ${trust} =~ [er] ]] && need_full=1
```
Why are you using regexps when pattern matching can do the same more readably and faster?

mgorny commented on autosign.bash:
```diff
@@ -171,47 +268,138 @@ sign_key() {
+	#
+	# The local user MUST be specified as a full fingerprint, for matching &
+	# security reasons.
+	trust_args=(
```
Global variables should be uppercase.

mgorny commented on autosign.bash:
```diff
@@ -171,47 +268,138 @@ sign_key() {
+			fpr) fpr=${GPGREC[10]} ;;
+		esac
+	done <"${f}"
+	case "$seen_sec" in
```
`${seen_sec}`

mgorny commented on autosign.bash:
```diff
@@ -171,47 +268,138 @@ sign_key() {
 	local k uid
 	# revoke signatures on old keys
-	while read uid k; do
+	while IFS=$(printf '\t') read uid k; do
```
`$'\t'`

mgorny commented on test.bash:
```diff
@@ -0,0 +1,140 @@
+# code picking a random key that was the correct one for the key.
+for f in $(seq 5 8) ; do
+	:
+	#gpg --batch --passphrase '' --quick-gen-key "Dummy Key <dummy-0x$(printf %04x $f)@example.com>"
```
Is this commented out intentionally?

mgorny commented on test.bash:
```diff
@@ -0,0 +1,140 @@
+	#gpg --batch --passphrase '' --quick-gen-key "Dummy Key <dummy-0x$(printf %04x $f)@example.com>"
+done
+
+if [[ -z "$authfpr_foo" ]] || [[ -z "$authfpr_bar" ]] || [ "$authfpr_bar" == "$authfpr_foo" ]; then
```
You can combine these into one `[[ ... ]]`.

mgorny reviewed (changes requested):
That said, it is horribly hard to read when you mix a dozen different changes into a single commit, then do some refactoring on top of them in the next one.

robbat2 committed:
feat: send to more keyservers

Signed-off-by: Robin H. Johnson <robbat2@gentoo.org>
